Implement Buffer Device Address for Rendering Device Vulkan and DirectX12

thimenesup opened the pull request:
This is an extremely useful feature that is supported on many GPUs which may be worthwhile to have it since the changes are minimally invasive, and allows for a lot of flexibility, specially for compute shaders and a requirement for ray tracing.

Tested to work on Windows 10 with Vulkan driver.
On DirectX12 the SPIR-V to NIR shader compiler doesn't have implemented this extension (yet?), so remains untested.
[Test project used available here](https://github.com/thimenesup/GodotBufferAddressExample)

Do note that it only has been implemented for storage buffers, to not change function signatures, but it could be implemented for the other buffer types, either by adding another default parameter, or breaking compatibility and using bitwise flags.

thimenesup committed:
Implement Buffer Device Address for Rendering Device Vulkan and DirectX12

DarioSamo commented:
Both you and https://github.com/godotengine/godot/pull/99119 are trying to implement this as buffer device address is required for RT as well. You might want to coordinate how to support this particular feature as both seem to be different. CC @Fahien

Fahien commented on servers/rendering/rendering_device.cpp:
```diff
@@ -645,6 +645,19 @@ Vector<uint8_t> RenderingDevice::buffer_get_data(RID p_buffer, uint32_t p_offset
+
+	Buffer *buffer = _get_buffer_from_owner(p_buffer);
+	if (!buffer) {
+		ERR_FAIL_V_MSG(0, "Buffer argument is not a valid buffer of any type.");
```
You can use `ERR_FAIL_COND_V_MSG(!buffer, 0, "Buffer ...");` on this line and below :)

thimenesup commented:
From what I have seen the implementations are fairly similar, albeit with some small differences, mine loads the extension either from core 1.2 or the KHR version which has a wider support compared to the EXT used in the other branch (sourced from the gpuinfo database), and also flags index and vertex buffers by default, which likely won't play nicely on devices where its not supported.

I think the buffer device address feature its worthwhile on its own and should be merged beforehand, then the other ray tracing branch could be rebased from it.

Went ahead an added it as optional functionality to index, vertex and uniform buffers while preserving compatibility, and updated the documentation.
Also implemented it for Metal to ensure compilation, but its an API I haven't really used, so I don't know for sure if its 100% correct.

DarioSamo commented:
> I think the buffer device address feature its worthwhile on its own and should be merged beforehand, then the other ray tracing branch could be rebased from it.

I believe so as well, it's easier to integrate it piece-meal like this. I just wanted to make sure you were both in agreement as it'll break the RT branch as soon as it's merged. I have no qualms with integrating the feature as I know how necessary it is for advanced rendering.

RandomShaper commented:
Looks good, but there are compatibility breakages.

thimenesup commented:
Added the GDExtension compatiblity methods and rebased with the current master again to fix merge conflicts.

dsnopek commented on core/extension/gdextension_compat_hashes.cpp:
```diff
@@ -698,14 +698,14 @@ void GDExtensionCompatHashes::initialize() {
-		{ "vertex_buffer_create", 3491282828, 3410049843 },
+		{ "vertex_buffer_create", 3491282828, 76995708 },
 		{ "vertex_array_create", 3137892244, 3799816279 },
-		{ "index_buffer_create", 975915977, 3935920523 },
+		{ "index_buffer_create", 975915977, 2360653086 },
 		{ "shader_compile_spirv_from_source", 3459523685, 1178973306 },
 		{ "shader_compile_binary_from_spirv", 1395027180, 134910450 },
 		{ "shader_create_from_spirv", 3297482566, 342949005 },
 		{ "shader_create_from_bytecode", 2078349841, 1687031350 },
-		{ "uniform_buffer_create", 1453158401, 34556762 },
+		{ "uniform_buffer_create", 1453158401, 3410049843 },
```
These changes should be removed. Registering the compatibility methods should be enough.

thimenesup replied:
Makes sense, I did suspect it when I saw the the file history didn't showcase any other hash changes

Fahien commented:
Do we need something like this in `_initialize_device()`?

```cpp
	VkPhysicalDeviceBufferDeviceAddressFeaturesKHR device_address_features = {};
	if (buffer_device_address_support) {
		device_address_features.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_BUFFER_DEVICE_ADDRESS_FEATURES_KHR;
		device_address_features.pNext = create_info_next;
		device_address_features.bufferDeviceAddress = buffer_device_address_support;
		create_info_next = &device_address_features;
	}
```

thimenesup commented:
I think it might be needed for the devices that aren't initializing core Vulkan 1.2 but a prior version instead.

stuartcarnie commented on drivers/metal/rendering_device_driver_metal.mm:
```diff
@@ -3947,6 +3956,8 @@ bool isArrayTexture(MTLTextureType p_type) {
 			return false;
 		case SUPPORTS_FRAGMENT_SHADER_WITH_ONLY_SIDE_EFFECTS:
 			return true;
+		case SUPPORTS_BUFFER_ADDRESS:
+			return false;
```
This can be true, but I would like to implement it by adding a new boolean property on the `MetalFeatures` struct:

https://github.com/godotengine/godot/blob/f27764ecec09469f13b25716f1adef59f1e63821/drivers/metal/metal_device_properties.h#L73

DarioSamo replied:
Could this be addressed in this PR?

stuartcarnie commented on drivers/metal/rendering_device_driver_metal.mm:
```diff
@@ -158,6 +158,15 @@ _FORCE_INLINE_ MTLSize mipmapLevelSizeFromSize(MTLSize p_size, NSUInteger p_leve
 	// Nothing to do.
 }
 
+uint64_t RenderingDeviceDriverMetal::buffer_get_device_address(BufferID p_buffer) {
+	if (@available(iOS 16.0, macOS 13.0, *)) {
+		id<MTLBuffer> obj = rid::get(p_buffer);
+		return obj.gpuAddress;
+	} else {
+		return 0;
+	}
+}
+
```
This is correct for Metal 👍🏻

stuartcarnie reviewed (commented):
I've left a couple of notes for Metal. Your implementation is correct, so I suggest you add a property to the `MetalFeatures` structure, that is set on initialisation, so you can return the correct value for the `SUPPORTS_BUFFER_ADDRESS` feature.

DarioSamo commented:
I'd like to push for this to be merged, but it seems there's some compatibility breakage that must be addressed. I'll give it a thorough review soon.

DarioSamo reviewed (changes requested):
So far this looks excellent to me, but I've noticed there's some inconsistency when it comes to choosing these terms:
- Device Address (`buffer_get_device_address`)
- Shader Address (`enable_shader_address`)
- Shader Device Address (`STORAGE_BUFFER_USAGE_SHADER_DEVICE_ADDRESS`)
- Buffer Address (`SUPPORTS_BUFFER_ADDRESS`)

If we can settle on one consistent terminlogy (device address sounds like the best one to me), and fix the conflicts on the API change, I think this is good to go.

```
buffer_get_device_address()
enable_device_address
STORAGE_BUFFER_USAGE_DEVICE_ADDRESS
SUPPORTS_BUFFER_DEVICE_ADDRESS
```

thimenesup commented:
Named the used terms consistently and rebased with master branch as requested.

DarioSamo commented:
Approved, thank you very much for your work!

clayjohn commented:
Updated with the changes requested by ATS. Should be ready to merge now

akien-mga commented:
Thanks!
